async-profiler-context/src/main/java/io/pyroscope/labels/Ref.java

```diff
@@ -16,13 +17,16 @@ public Ref(T val, Long id) {
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
-        Ref<T> valueRef = (Ref<T>) o;
-        return id.equals(valueRef.id);
+
+        Ref<?> valueRef = (Ref<?>) o;
+        if (val.getClass() != o.getClass()) return false;
+
+        return id.equals(valueRef.id) && val.equals(valueRef.val);
     }
 
     @Override
     public int hashCode() {
-        return id.hashCode();
+        return Objects.hash(id, val);
```

Comment:
I am not sure we want this change. The value here can be a string or a container type (we seem to be using maps as T), and if it is a container type we'll end up doing a deep traversal for `equals()` and `hashCode()` as opposed to a simple integer comparison / hashing which can become bad for performance.

Semantically, ids should be unique and generated based on a similar (and somewhat expensive) [map lookup](https://github.com/grafana/pyroscope-java/blob/1cc00a1693692359e5ef45c1ebcd39ac1573e22a/async-profiler-context/src/main/java/io/pyroscope/labels/RefCounted.java#L38-L39), so we should be fine with using the id only.

@korniltsev can you confirm this?

Reply:
In https://github.com/grafana/pyroscope-java/pull/175/commits/32f2f7771b61edd3e06f92f9b7fff672b1d09dac I reverted this change. I did this in the interest of keeping the changes in this PR non-controversial. I mostly wanted this PR to clean up a few untidy spots in the code base.

Reply:
> @korniltsev can you confirm this?

I confirm the id already represents the contents of the value which is uniquely stored in a mp Aleks pointed. It will not hurt to hash/equals the value, but it is also not needed.